### DISCUSSION TITLE
PDF DPI for thumbnail

### DISCUSSION THREAD

**Discussion by andrews05** [libvips/libvips#2466] [Q&A] [answered]
## Bug report

**Describe the bug**
When creating a thumbnail of a pdf with size 'down', the resulting image dimensions may be smaller than expected.

**To Reproduce**
1. Use attached pdf.
2. Run `vipsthumbnail -s '480x480>' gamecards.pdf`.

**Expected behavior**
The resulting thumbnail dimensions should be 450x480.

**Actual behavior**
The resulting thumbnail dimensions are 425x454.

**Screenshots**

[gamecards.pdf](https://github.com/libvips/libvips/files/7297497/gamecards.pdf)

**Environment**
 - OS: Ubuntu 20.04
 - Vips: 8.11.4

**Additional context**
The problem is presumed to be a result of loading the PDF with a DPI of 72, making vips think the input size is 425x454. I believe using a DPI of at least 150 would be preferrable.
Obviously, I could avoid the issue by *not* using size 'down', it's just that I'm using the same code path for all image types and I don't want raster images to be scaled up.

**Comment by jcupitt** [comment, marked as answer]
Hi @andrews05,

I think this is by design: vector images have a natural size, and if you ban upsizing they just will be too small. SVG has the same issue.

You can override the default 72 DPI like this:

```
$ vipsthumbnail -s '480x480>' gamecards.pdf[dpi=150] --vips-info
VIPS-INFO: 04:36:11.056: thumbnailing gamecards.pdf[dpi=150]
VIPS-INFO: 04:36:11.058: selected loader is VipsForeignLoadPdfFile
VIPS-INFO: 04:36:11.058: input size is 886 x 945
VIPS-INFO: 04:36:11.058: loading with factor 1.96875 pre-shrink
VIPS-INFO: 04:36:11.058: pre-shrunk size is 450 x 480
VIPS-INFO: 04:36:11.058: converting to processing space srgb
VIPS-INFO: 04:36:11.058: thumbnailing gamecards.pdf[dpi=150] as ./tn_gamecards.jpg
```

You can see it's now taking 886x945 as the natural size, then reloading with a scale that gets you to 450x480. You can set any DPI:

```
$ vipsthumbnail -s '480x480>' gamecards.pdf[dpi=10000] --vips-info
VIPS-INFO: 04:39:07.512: thumbnailing gamecards.pdf[dpi=10000]
VIPS-INFO: 04:39:07.514: selected loader is VipsForeignLoadPdfFile
VIPS-INFO: 04:39:07.514: input size is 59055 x 62992
VIPS-INFO: 04:39:07.514: loading with factor 131.233 pre-shrink
VIPS-INFO: 04:39:07.515: pre-shrunk size is 450 x 480
VIPS-INFO: 04:39:07.515: converting to processing space srgb
VIPS-INFO: 04:39:07.515: thumbnailing gamecards.pdf[dpi=10000] as ./tn_gamecards.jpg
```

And there's no performance cost, since it never renders at that resolution. The same trick works for SVG.

**Reply by andrews05** [reply]
Perfect, thank you! I’ll check how to do this in php…

**Reply by jcupitt** [reply]
Are you using php-vips? You need to append the `[dpi=150]` to the filename, but only for PDF and SVG files.

You can sniff the filetype by checking the `vips-loader` metadata item.
